Dev-Corner/tutorials/visualizations.md

```diff
@@ -39,73 +49,527 @@ Documentation about Galaxy visualizations can be found here. It can be used as b
+files does not take place on the server, but instead data will be downloaded by the client via an
+exposed Galaxy URL prior to client-side rendering.
+
+The most basic part of the mako file are the variables used for further web development, given
+below.
+
+```mako
+<!DOCTYPE HTML>
+<%
+    ## Generates hash (hdadict['id']) of history item
+    hdadict = trans.security.encode_dict_ids( hda.to_dict() )
+
+    ## Finds the parent directory of galaxy (/, /galaxy, etc.)
+    root     = h.url_for( '/' )
+
+    ## Determines the exposed URL of the ./static directory
+    app_root = root + 'plugins/visualizations/'+visualization_name+'/static/'
+
+    ## Actual file URL:
+    file_url = root + 'datasets/' + hdadict['id'] + "/display?to_ext=" + hda.ext;
+%>
+```
+
+- `hdadict` is a variable that contains a file identifier that has been encoded to it's
+   exposed uid.
+- `root` indicates location of Galaxy on the webserver (e.g. `/`, `/galaxy/`, `/galaxy-pub/`, etc).
+- `app_root` contains the exposed url of the static files for this visualization.
+- `file_url` contains the exposed url of the dataset selected (by user) for visualization.
+
+We could obtain the BAM file client-side by downloading the BAM/SAM file in its entirety via
+*file_url*. However, BAM files can become quite large and it is usually not desired to transfer
+such datasets over the network. In our case it is also rather inconvenient to parse the BAM file
+with Javascript, just to count the number of reads.
+
+Fortunately, BAM files have indices. These indices are brief summaries describing the
+number of entries per chromosome, in order to be able to access the data contained in them more
+quickly. In the mako template we can access a BAM index as *hda.metadata.bam_index*.
+(Note that this is a file path on the server, not an exposed URL).
+
+Samtools has a command named `idxstats` which is able to leverage this BAM index you. However,
+since visualizations do not have dependency management, it is very tricky to let the mako template
+do a system call to samtools. Fortunately, the Galaxy ecosystem ships with a built-in `pysam`
+dependency, a library that can do any native samtools command within python.
+
+The `*.metadata.bam_index` is a special kind of file in the Galaxy ecosystem. It is actually an
+invisible file in Galaxy, linked to another history item, but does have a unique filename.
+So, for the BAM file `./database/files/000/dataset_001.dat`, our BAI file (index) is **not**
+`./database/files/000/dataset_001.dat.bai` but could be `./database/files/000/dataset_002.dat` or
+`./database/files/000/dataset_003.dat`. We can create a symlink to this index file to ensure the
+bam file and its index share the same prefix, as expected by samtools and pysam.
+
+We can create this symlink as follows in our mako template:
+
+```python
+## Ensure BAI index is symlinked
+bai_target = hda.file_name+'.bai'
```

Comment:
This will pollute the Galaxy DB, isn't it?
A better way would be to link both file into a temp_dir created by `tempfile`